actors/builtin/miner/miner_actor.go

```diff
@@ -160,53 +173,166 @@ func (a Actor) ChangePeerID(rt Runtime, params *ChangePeerIDParams) *adt.EmptyVa
+
+		// TODO WPOST (follow-up): process Skipped as faults
 
-		// increment proving period start
-		// Note: this must happen after verifyWindowedPoSt, lest verification use the wrong randomness
-		// (drawn from ProvingPeriodStart)
-		st.PoStState = PoStState{
-			ProvingPeriodStart:     st.PoStState.ProvingPeriodStart + ProvingPeriod,
-			NumConsecutiveFailures: 0,
+		// Extract a fault set relevant to the sectors being submitted, for expansion into a map.
+		declaredFaults, err := bitfield.IntersectBitField(provenSectors, st.Faults)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to intersect proof sectors with faults: %v", err)
+		}
+		declaredRecoveries, err := bitfield.IntersectBitField(declaredFaults, st.Recoveries)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to intersect recoveries with faults: %v", err)
+		}
+		expectedFaults, err := bitfield.SubtractBitField(declaredFaults, declaredRecoveries)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to subtract recoveries from faults: %v", err)
 		}
 
-		// reset provingSet to include all sectors (were not included during challenge period)
-		st.ProvingSet = st.Sectors
+		nonFaults, err := bitfield.SubtractBitField(provenSectors, expectedFaults)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to diff bitfields: %s", err)
+		}
 
+		empty, err := nonFaults.IsEmpty()
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to check if bitfield was empty: %s", err)
+		}
+		if empty {
+			rt.Abortf(exitcode.ErrIllegalArgument, "no non-faulty sectors in partitions %s", params.Partitions)
+		}
+
+		// Select a non-faulty sector as a substitute for faulty ones.
+		goodSectorNo, err := nonFaults.First()
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to get first good sector: %v", err)
+		}
+
+		// Load sector infos for proof
+		sectorInfos, err := st.LoadSectorInfosWithFaultMask(store, provenSectors, expectedFaults, abi.SectorNumber(goodSectorNo))
+		rt.Abortf(exitcode.ErrIllegalState, "failed to load sector infos: %v", err)
+
+		// Verify the proof.
+		// A failed verification doesn't immediately cause a penalty; the miner can try again.
+		a.verifyWindowedPost(rt, challengeEpoch, sectorInfos, params.Proofs)
+
+		// Record the successful submission
+		postedPartitions := bitfield.NewFromSet(params.Partitions)
+		contains, err := abi.BitFieldContainsAny(st.PostSubmissions, postedPartitions)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to intersect post partitions")
+		if contains {
+			rt.Abortf(exitcode.ErrIllegalArgument, "duplicate PoSt partition")
+		}
+		err = st.AddPoStSubmissions(postedPartitions)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to record submissions for partitions %s: %v", params.Partitions, err)
+		}
+
+		// If the PoSt was successful, the declared recoveries should be restored
+		err = st.RemoveFaults(store, declaredRecoveries)
```

Comment:
What happens if a sector was not declared as recovered but gets proven in this windowPost?

It must not be accounted for power until the next windowpost

Reply:
The PoSt will fail validation. The good sectors + declared recoveries must exactly match the set proven.